### DISCUSSION TITLE
Generated PreviewModels swift file has no access to preview files from Swift local packages

### DISCUSSION THREAD

**Discussion by woutergoossens** [BarredEwe/Prefire#32] [Ideas]
## Context 🕵️‍♀️
We have rather big project with some local SPM local packages. In those packages we have SwiftUI screens and components. 
I added Prefire to the project and package and updated a SwiftUI preview in the local package to use PrefireProvider.
But then the project generated the PreviewModels file but it has no access to  

<img width="1083" alt="Screenshot 2023-12-06 at 16 37 28" src="https://github.com/BarredEwe/Prefire/assets/355048/966e4d17-07d1-46d0-9201-6f12c403682e">

AppointmentDetailsScreen_Previews is part of Presentation package and that package isn't imported or cannot be accessed.. How can I do that?
Can I update the PreviewModels template file?

**Comment by woutergoossens** [comment]
It seems PR 21 is a possible solution? Any ETA on this?
https://github.com/BarredEwe/Prefire/pull/21

**Comment by BarredEwe** [comment]
I'm currently working on a new major release https://github.com/BarredEwe/Prefire/pull/29 with a lot of improvements. In particular, the ability to install custom imports has been added:

```
test_configuration:
  - target: PrefireExample
  - simulator_device: "iPhone15,2"
  - required_os: 17

playbook_configuration:
  - imports:
      - UIKit
      - Foundation
```

⚠️ The branch is currently under active development, but the main work has already been done and the general interface has already been fixed.

**Comment by woutergoossens** [comment]
Sounds great! 
Can I already test this via SPM with your specific branch? 

Our packages also have tests, is that supported? So we run the test in the main project and snapshot tests from the packages are verified? 

So in test_configuration target would be just our main target?

When do you think it could be released?
I am working for a big Belgian banking company and we have more then > 100 screens and components in SwiftUi that we like to have snapshots from (and tested) :)

so would be a good usecase to test v2, right?

**Comment by BarredEwe** [comment]
I think it's really possible to use a branch now. 
According to the plans, there is not much left to finish. About 1-2 weeks before the full release. 
In general, it sounds great for a test and a quick revision :)

**Comment by woutergoossens** [comment]
Sounds good. Happy to test :) 

I added that branch via SPM but when I run the tests I get this

<img width="1207" alt="Screenshot 2023-12-06 at 20 47 37" src="https://github.com/BarredEwe/Prefire/assets/355048/82195645-907d-4507-8468-ff62368fc197">

**Comment by BarredEwe** [comment]
I can't reproduce it ):
Could you clear spm cache?

**Comment by woutergoossens** [comment]
Tried cleaning everything.. Same issue.. 
Will try in the example project later today...

**Comment by woutergoossens** [comment]
I have the same issue when I run the example app.. Let me try again clearing everything :D 

Nope.. doesn't work. Should I do something extra? 

<img width="947" alt="Screenshot 2023-12-07 at 11 00 12" src="https://github.com/BarredEwe/Prefire/assets/355048/1aca207f-d4cd-4a36-971f-55bd073b2527">

**Comment by BarredEwe** [comment]
I've added a couple of changes, can you check on the new version of the branch?

**Comment by woutergoossens** [comment]
Thanks. That works :) I added it to my project and a package with SwiftUI screens. 
But now I have an issue with the new macro previews.

For example it wants to generate the PreviewTests.generated file like this -->
<img width="1144" alt="Screenshot 2023-12-08 at 09 23 54" src="https://github.com/BarredEwe/Prefire/assets/355048/3e6ad7da-2a40-4a89-aad6-ccbf6df4352f">

We create that screen with a private PreviewViewModel (the VM is our ObservableObject with state and events,...)
But the generated file has no access to it.. Also we add an environmentObject to it, which is part of another package. How can I add that import for the tests? How can I fix that without making all those classes public? :) 

In the old way of creating previews, it's also created like this and that works without making the VM public.

```
static var previews: some View {
        SplashScreen(viewModel: PreviewViewModel())
    }
```

**Comment by woutergoossens** [comment]
So.. tested now in the Project itself.

So in our project we have SwiftUI files in different packages (for example Presentation, DesignSwiftUI,...)

In the custom config I added extra import Presentation

```
playbook_configuration:
  - imports:
      - UIKit
      - Foundation
      - Presentation
```

PreviewModels -->
So that is working. The file is found. 

* BUT the SplashScreen_Previews struct should be public, otherwise it doesn't find the file.. 
* Macro previews are not showed?

<img width="943" alt="Screenshot 2023-12-08 at 13 26 47" src="https://github.com/BarredEwe/Prefire/assets/355048/18fd8540-d585-4092-a080-2cafbdd3156d">

PreviewTests -->
Macro Previews:
* I had to remove the private from the PreviewViewModel (see previous comment).
* I get Invalid redeclaration of tests.. Those previews are only declared once.. Already cleaned all..
It seems it cached the tests and doesn't remove/update them when I update the preview declaration?
For example I changed/renamed the environmentObject but it keeps track of it 

<img width="1084" alt="Screenshot 2023-12-08 at 13 31 00" src="https://github.com/BarredEwe/Prefire/assets/355048/604302a7-f31f-41b4-9a20-9b5b6225f043">

<img width="992" alt="Screenshot 2023-12-08 at 15 34 54" src="https://github.com/BarredEwe/Prefire/assets/355048/8c7a19e7-133d-48df-b63c-53bb83f9715a">

**Comment by BarredEwe** [comment]
I have added a new configuration:
```yaml
prefire_configuration:
  - imports:
    - UIKit
    - Foundation
  - testable_imports:
    - SwiftUI
```

And I fixed a few bugs.

**Comment by woutergoossens** [comment]
Thanks! 

Playbook (PreviewModels.generated) is working now with @testable import Presentation
But doesn't show the Macro previews? 

PreviewTests is still failing because of Invalid redeclaration of 'test_SelectCoHolderTestPreview_Preview()'
Already tried to remove the SwiftUI file but it keeps the test..
<img width="582" alt="Screenshot 2023-12-11 at 14 16 30" src="https://github.com/BarredEwe/Prefire/assets/355048/da77566f-5483-45a2-9cfa-fc050d7622cf">

What can I do to fix that? Already cleaned DerrivedData/SPM cache/...

**Comment by BarredEwe** [comment]
I hope I fixed it)

**Comment by woutergoossens** [comment]
Worked once, then it was cleaned and the method was only there once.
Then I did a change. Changed the preview macro to old PreviewProvider and the old macro stays in the PreviewTests file..

Do you need some logs or more info?

**Comment by woutergoossens** [comment]
Let me try to reproduce it with the example app :)

**Comment by woutergoossens** [comment]
Hello, while trying to reproduce this I got this weird issue. 
PreviewTests is combining tests of the Example project and my project.. :D 
It's mixing some saved cache?

<img width="1327" alt="Screenshot 2023-12-13 at 12 25 39" src="https://github.com/BarredEwe/Prefire/assets/355048/16ab4f67-e5f4-4320-b1ec-5369ad77d0e0">

**Comment by BarredEwe** [comment]
Everything that is generated by macros is in one folder. Now the filtering is very simple (search for the inclusion of the project name in the file name). I'll try to figure out how to improve it.

**Comment by BarredEwe** [comment]
I have added a new macro file filtering. Macros from another project should no longer be included

**Comment by woutergoossens** [comment]
Thanks! That's working.

But I still have an issue ==> I need multiple testable imports 

Our structure:
AppTests target will run the tests from the App target + tests from all the Swift Packages (and we need tests from Presentation, DesignSwiftUI,...)

How can we have that?
For example now I have as config:

```
test_configuration:
  - target: Presentation
```

And in the generated PreviewTests file it is combining the test from AppTests target + packages. So it needs to have multiple @testable import statements. 

For example the generated file now has:

```
@testable import Presentation
 private var file: StaticString { .init(stringLiteral: "/Users/gowo/Development/**/mobile-ios/Presentation/PreviewTests.swift") }

```
But in my case it could better be 

```
@testable import Presentation
@testable import DesignSwiftUI

 private var file: StaticString { .init(stringLiteral: "/Users/gowo/Development/**/mobile-ios/AppTests/PreviewTests.swift") }

```

because now it finds the previews from the Presentation package but it also generated the preview tests for the DesignSwiftUI package but it doesn't find those files :) 

Thanks!!

**Comment by BarredEwe** [comment]
Added parameter setting for test_configuration (same as for playbook_configuration):
```yaml
test_configuration:
  - required_os: 17
  - imports:
      - UIKit
      - SwiftUI
  - testable_imports:
      - Prefire
```

**Comment by woutergoossens** [comment]
Thanks! It's working!! 🥳 

But I think I found an issue regarding the Macro's.

I updated the test_configuration target to AppName_STUB_Debug as this is the main App module for the tests. And so it would create the snapshots in the AppName_STUB_Debug folder, instead of the Presentation folder.
<img width="781" alt="Screenshot 2023-12-15 at 15 36 43" src="https://github.com/BarredEwe/Prefire/assets/355048/f07aad32-3e10-4890-bf5a-7f4638f82905">

So I have 2 testable imports (via the config) + 1 import added via the target

@testable import Argenta_STUB_Debug
@testable import Presentation
@testable import DesignSwiftUI

But when I do that, the macro previews are not included in the PreviewTests file, only the ones from PreviewProvider.
When I update the target to DesignSwiftUI or Presentation, the macro previews from the chosen package are included.

` private var file: StaticString { .init(stringLiteral: "/Users/gowo/Development/**/mobile-ios/AppName_STUB_Debug/PreviewTests.swift") }`

How can we fix that? :) 

Almost there! Thanks for the great work and effort! 

EDIT: I think we should combine all the preview bodies, not only from specified target but also from the testable import targets?

**Comment by woutergoossens** [comment]
And I tried to build it via our CI and got this issue

 error: “PrefirePlaybookPlugin” must be enabled before it can be used

On Xcode I get a popup to trust the plugin.. But hmm how should we do that on CI? :)

**Comment by woutergoossens** [comment]
> And I tried to build it via our CI and got this issue
> 
> error: “PrefirePlaybookPlugin” must be enabled before it can be used
> 
> On Xcode I get a popup to trust the plugin.. But hmm how should we do that on CI? :)

I think I found it.
I should use `defaults write com.apple.dt.Xcode IDESkipPackagePluginFingerprintValidatation -bool YES`

**Comment by BarredEwe** [comment]
> Thanks! It's working!! 🥳
> 
> But I think I found an issue regarding the Macro's.
> 
> I updated the test_configuration target to AppName_STUB_Debug as this is the main App module for the tests. And so it would create the snapshots in the AppName_STUB_Debug folder, instead of the Presentation folder. <img alt="Screenshot 2023-12-15 at 15 36 43" width="781" src="https://private-user-images.githubusercontent.com/355048/290851609-f07aad32-3e10-4890-bf5a-7f4638f82905.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI3NDk5MTQsIm5iZiI6MTcwMjc0OTYxNCwicGF0aCI6Ii8zNTUwNDgvMjkwODUxNjA5LWYwN2FhZDMyLTNlMTAtNDg5MC1iZjVhLTdmNDYzOGY4MjkwNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxNlQxODAwMTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03MjA4ZjdmNTBkZjI4MjBjODRkNzM1ZjE3ZTVkOTljZDg1NzBhYmQ0NjgxYzlmNjgyMDBkNDc3MTNjODcxNTY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6LO0H4A9MD2XoFfUAnM1j2elSRN1QvH7Y83Kys80g3g">
> 
> So I have 2 testable imports (via the config) + 1 import added via the target
> 
> @testable import Argenta_STUB_Debug @testable import Presentation @testable import DesignSwiftUI
> 
> But when I do that, the macro previews are not included in the PreviewTests file, only the ones from PreviewProvider. When I update the target to DesignSwiftUI or Presentation, the macro previews from the chosen package are included.
> 
> ` private var file: StaticString { .init(stringLiteral: "/Users/gowo/Development/**/mobile-ios/AppName_STUB_Debug/PreviewTests.swift") }`
> 
> How can we fix that? :)
> 
> Almost there! Thanks for the great work and effort!
> 
> EDIT: I think we should combine all the preview bodies, not only from specified target but also from the testable import targets?

I understood the problem. I've redesigned the macro filtering system. Now it will enable all macros along the path of the main target.

**Comment by woutergoossens** [comment]
Hi,

Thanks for the update. But now it doesn't include any macro previews.. Should I update the config?
Only when I change the target name to a package name, I see the macros..

Our app folder structure. Every package has a folder in the root. 

/AppName 
/Presentation
/DesignSwiftUI
/... 

And our main Target is something like AppName_STUB_Debug which hasn't a folder in this stucture.
AppName_STUB_Debug folder is only created to put the snapshots.. 

When I remove target from the config it will use "App" target. But then I get this -->
<img width="1213" alt="Screenshot 2023-12-18 at 15 28 28" src="https://github.com/BarredEwe/Prefire/assets/355048/f7d0bfa5-a6e1-4748-a42f-340f0c24e459">

App is the target name but `AppName_STUB_Debug` is our product/module name for testing..

Almost there! How can I make this work? 

Wouter

**Comment by woutergoossens** [comment]
Let me try create a sample app :)

**Comment by woutergoossens** [comment]
So I created an example app with 2 Swift Packages.

Presentation has an AuthView and Design has PrefireView

But here also have an issue ==>
When it creates the PreviewTests file the first time, it contains the preview macro (but only 1 per View). 
But when I add some more macro previews, the PreviewTests file is not updated.. It only adds the first one from a file?

Good thing is that it adds the preview macro's from the packages

[PrefireExample.zip](https://github.com/BarredEwe/Prefire/files/13740447/PrefireExample.zip)

**Comment by BarredEwe** [comment]
Thanks for the example project! I found the problem and posted a fix.

**Comment by woutergoossens** [comment]
Thanks. Working great now. It generated 300 snapshots now of our SwiftUI Screens :) 

What is the progress on the 2.0?

**Comment by BarredEwe** [comment]
Version 2.0 is almost ready. It remains to complete the tests and documentation)

**Comment by woutergoossens** [comment]
@BarredEwe Hi, thanks for version 2!

Works great locally but having issues to run it on our CI.

What am I doing wrong? I also want to look for an iphoneos folder instead of the iphonesimulator folder? 

Always get this issue
`Error: error: '/Users/***/actions-runner/_work/mobile-ios/mobile-ios/DerivedData/stub/Build/Intermediates.noindex/ArchiveIntermediates/Argenta stub/BuildProductsPath/STUB-Debug-iphoneos/Argenta.app' does not exist or is not readable.
** TEST INTERRUPTED **
The following build commands failed:
	Apply build tool plug-in “PrefireTestsPlugin” to target “AppTests” in project “Argenta”
	eyJ0eXBlIjp7IndvcmtzcGFjZSI6e319fQ==`

**Reply by woutergoossens** [reply]
Our CI is using fastlane scan to run tests.

We also see this issue now..
So related to the plugin..

We use a custom derived_data_path https://docs.fastlane.tools/actions/run_tests/ - maybe that is the issue? 

```

❌  error: '/Users/gowo/Development/Argenta/mobile-ios/SourcePackages/checkouts/Nimble-SnapshotTesting/_Pods.xcodeproj' does not exist or is not readable.


** TEST INTERRUPTED **


The following build commands failed:
	Apply build tool plug-in “PrefireTestsPlugin” to target “AppTests” in project “Argenta”
	eyJ0eXBlIjp7IndvcmtzcGFjZSI6e319fQ==
```

**Comment by woutergoossens** [comment]
I updated the example app with Fastlane support 

and it seems the generated file does not use the .prefire.yml file? It doesn't contained the testable imports..
[PreviewTests.generated.swift.zip](https://github.com/BarredEwe/Prefire/files/14335700/PreviewTests.generated.swift.zip)

Full project with Fastlane.
Run `bundle exec fastlane test`       
[PrefireExample.zip](https://github.com/BarredEwe/Prefire/files/14335708/PrefireExample.zip)

```
Testing failed:
	Cannot find 'DesignPrefireView_Preview' in scope
	Cannot find 'PresentationAuthView_Preview' in scope
	Cannot find 'DesignPrefireView' in scope
	Cannot find 'DesignPrefireView' in scope
	Cannot find 'PresentationAuthView' in scope
	Cannot find 'DesignPrefireView' in scope
	Cannot find 'PresentationAuthView' in scope
	Testing cancelled because the build failed.
```

Running it in Xcode works perfect..

**Comment by woutergoossens** [comment]
I think I found an initial problem. When we use Fastlane, we fetch SPM packages in a SourcePackages folder + we use a custom DerivedData folder in the root of the project. So we can clean/cache those afterwards. 
The `Argenta.app` in `DerrivedData/*...` and `Nimble-SnapshotTesting/_Pods.xcodeproj`are aliases (that are not readable or corrupted it seems). 

But the build tool plugin has issues with those files. When I update the paths (outside of the root of the project or just the default ones) it continues to build and then I have the same issue as in the example test project. The generated file does not seems to use the .prefile.yml config file (so target and testable_imports are not used..) and the test build fails.

Any idea how to make this work via Fastlane?

**Comment by woutergoossens** [comment]
Extra log:
<img width="1521" alt="Screenshot 2024-02-21 at 08 56 24" src="https://github.com/BarredEwe/Prefire/assets/355048/8dbdc814-849d-409b-b810-ed554c3fb2fb">

So it looks in another DerivedData folder (in the root of the project). 
Having a list of excluded folders would be the easiest I think?

**Comment by woutergoossens** [comment]
[_Pods.xcodeproj.zip](https://github.com/BarredEwe/Prefire/files/14356272/_Pods.xcodeproj.zip)

If you add this file (unzipped) somewhere in the root, you will get the error.

**Reply by BarredEwe** [reply]
I'm not sure yet, but maybe this fix can help: https://github.com/BarredEwe/Prefire/discussions/35#discussioncomment-8540269

**Comment by woutergoossens** [comment]
@BarredEwe Found a solution for my initial problem (when I run the tests via fastlane it did not find the .prefire.yml file (or didn't use it))

==> Via Fastlane I was running the tests via my `App` scheme but to make it work we need to run the `AppTests` scheme and move the .prefire.yml file to the AppTests folder. And then it works.. 

So only "issue" that remains is defining the correct Source folder or excluding a folder. 

Thanks!
